tests/utils/tensorflow/test_model_data_utils.py

```diff
@@ -22,14 +23,225 @@
+from rasa.utils.tensorflow.constants import SENTENCE, SEQUENCE
 from rasa.shared.nlu.training_data.message import Message
 from rasa.shared.nlu.training_data.training_data import TrainingData
 from rasa.utils.tensorflow.model_data_utils import TAG_ID_ORIGIN
 
 shape = 100
 
 
+@pytest.mark.parametrize("seed", [123, 234])
+def test_combine_attribute_features_from_message_filters_correctly(seed: int,):
+    """Checks that we extract and combine the right features.
+
+    We check that we correctly filter
+    (1) which feature goes under which key in the output dictionary according to
+        whether it `is_sparse` and according to it's `type` and
+    (2) which features should be included according to the requested attribute and
+        list of featurizers,
+
+    Note that for (2) we create attributes and origin information at random such that
+    there are exactly two attributes and two featurizers used overall and we
+    request exactly of the two attributes and one of the two origins, respectively.
+    """
+    rng = np.random.default_rng(seed=seed)
+
+    # the total number of `Features` we want to create
+    sparse_type_combinations = list(
+        itertools.product([True, False], [SEQUENCE, SENTENCE])
+    )
+    num_feats_per_sparse_type_combination = 4
+    total_number_of_features = (
+        len(sparse_type_combinations) * num_feats_per_sparse_type_combination
+    )
+
+    # create 2 dummy attributes and featurizers ("origin")
+    attribute_flag = rng.choice(2, size=total_number_of_features)
+    featurizer_flag = rng.choice(2, size=total_number_of_features)
+    for flags in [attribute_flag, featurizer_flag]:
+        assert (
+            sum(flags) > 0 and sum(flags) < total_number_of_features
+        ), "Fix this test. Choose a different seed."
+
+    # choose different dimensions for sparse and dense features (bool = is_sparse)
+    last_dimensions = {True: 3, False: 4}
+
+    # create dummy features
+    message_features = []
+    feat_idx = 1  # ensure each feature has uniue values
+    for idx, (is_sparse, feature_type) in enumerate(sparse_type_combinations):
+        seq_len = 1 if feature_type == SENTENCE else 4
+        feat_shape = (seq_len, last_dimensions[is_sparse])
+        for _ in range(num_feats_per_sparse_type_combination):
+            feat_data = (
+                np.eye(*feat_shape) if not is_sparse else scipy.sparse.eye(*feat_shape)
+            )  # shape doesn't matter, should be checked/enforced by `Feature`
+            feat_data = feat_data * feat_idx
+            feat_idx += 1
+            feat = Features(
+                feat_data,
+                feature_type=feature_type,
+                attribute=str(attribute_flag[idx]),
+                origin=str(featurizer_flag[idx]),
+            )
+            message_features.append(feat)
+
+    # shuffle them
+    permutation = rng.permutation(len(message_features))
+    message_features = [message_features[idx] for idx in permutation]

```

Comment:
Not sure if this works with any content, but maybe `np.random.shuffle(message_features)` does the same thing?

Reply:
oh, nice, that does work with arbitrary mutable sequences indeed :)